app/src/main/java/com/duckduckgo/app/bookmarks/ui/BookmarkAddEditDialogFragment.kt

```diff
@@ -0,0 +1,154 @@
+    }
+
+
+    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+
+        val rootView = View.inflate(activity, R.layout.add_or_edit_bookmark, null)
+        val titleInput = rootView.find<EditText>(R.id.titleInput)
+        val urlInput = rootView.find<EditText>(R.id.urlInput)
+
+        val alertBuilder = AlertDialog.Builder(activity!!)
+                .setView(rootView)
+                .setPositiveButton(R.string.bookmarkSave) { _, _ ->
+                    userAcceptedDialog(titleInput, urlInput)
+                }
+
+        validateBundleArguments()
+
+        setAlertTitle(alertBuilder)
+        populateFields(titleInput, urlInput)
+
+        val alert = alertBuilder.create()
+        showKeyboard(titleInput, alert)
+        return alert
+    }
+
+    private fun userAcceptedDialog(titleInput: EditText, urlInput: EditText) {
+        if (isInEditMode()) {
+            val listener = activity as BookmarkDialogEditListener
+            listener.userWantsToEditBookmark(getExistingId(),
+                    titleInput.text.toString(),
+                    urlInput.text.toString()
+            )
+        } else {
+            val listener = activity as BookmarkDialogCreationListener
+            listener.userWantsToCreateBookmark(
+                    titleInput.text.toString(),
+                    urlInput.text.toString())
+        }
+    }
+
+    private fun showKeyboard(
+            titleInput: EditText,
+            alert: AlertDialog
+    ) {
+        titleInput.setSelection(titleInput.text.length)
+        titleInput.showKeyboard()
+        alert.window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE)
+    }
+
+    private fun populateFields(titleInput: EditText, urlInput: EditText) {
+        titleInput.setText(getExistingTitle())
+        urlInput.setText(getExistingUrl())
+    }
+
+    private fun setAlertTitle(alertBuilder: AlertDialog.Builder) {
+        val titleStringRes = if (isInEditMode()) {
+            R.string.bookmarkTitleEdit
+        } else {
+            R.string.bookmarkTitleSave
+        }
+        alertBuilder.setTitle(titleStringRes)
+    }
+
+    private fun getExistingId(): Int = arguments!!.getInt(KEY_BOOKMARK_ID)
+    private fun getExistingTitle(): String = arguments!!.getString(KEY_PREEXISTING_TITLE)
+    private fun getExistingUrl(): String = arguments!!.getString(KEY_PREEXISTING_URL)
+
+    private fun validateBundleArguments() {
+        if (arguments == null) throw IllegalArgumentException("Missing arguments bundle")
+        val args = arguments!!
+        if (!args.containsKey(KEY_IS_EDIT_MODE) || !args.containsKey(KEY_PREEXISTING_TITLE) || !args.containsKey(
+                        KEY_PREEXISTING_URL
+                )) {
+            throw IllegalArgumentException("Bundle arguments required [KEY_IS_EDIT_MODE, KEY_PREEXISTING_TITLE, KEY_PREEXISTING_URL]")
+        }
+    }
+
+    private fun isInEditMode(): Boolean = arguments!!.getBoolean(KEY_IS_EDIT_MODE)
```

Comment:
This troubled me because it uses the unwraps the arguments but they are validated elsewhere first so no worries.